tensorflow/python/data/kernel_tests/matching_files_dataset_op_test.py

```diff
@@ -0,0 +1,267 @@
+        sess.run(itr.get_next())
+
+  def testSimpleDirectoryInitializer(self):
+    filenames = ['a', 'b', 'c']
+    self._touchTempFiles(filenames)
+
+    filename_placeholder = array_ops.placeholder(dtypes.string, shape=[])
+    dataset = MatchingFilesDataset(filename_placeholder)
+
+    with self.cached_session() as sess:
+      itr = iterator_ops.Iterator.from_structure(dataset.output_types)
+      init_op = itr.make_initializer(dataset)
+      next_element = itr.get_next()
+      sess.run(
+        init_op,
+        feed_dict={filename_placeholder: path.join(self.tmp_dir, '*')})
+
+      full_filenames = []
+      produced_filenames = []
+      for filename in filenames:
+        full_filenames.append(
+          compat.as_bytes(path.join(self.tmp_dir, filename)))
+        produced_filenames.append(compat.as_bytes(sess.run(next_element)))
+
+      self.assertItemsEqual(full_filenames, produced_filenames)
+
+      with self.assertRaises(errors.OutOfRangeError):
+        sess.run(itr.get_next())
+
+  def testFileSuffixes(self):
+    filenames = ['a.txt', 'b.py', 'c.py', 'd.pyc']
+    self._touchTempFiles(filenames)
+
+    filename_placeholder = array_ops.placeholder(dtypes.string, shape=[])
+    dataset = MatchingFilesDataset(filename_placeholder)
+
+    with self.cached_session() as sess:
+      itr = iterator_ops.Iterator.from_structure(dataset.output_types)
+      init_op = itr.make_initializer(dataset)
+      next_element = itr.get_next()
+      sess.run(
+        init_op,
+        feed_dict={filename_placeholder: path.join(self.tmp_dir, '*.py')})
+
+      full_filenames = []
+      produced_filenames = []
+      for filename in filenames[1:-1]:
+        full_filenames.append(
+          compat.as_bytes(path.join(self.tmp_dir, filename)))
+        produced_filenames.append(compat.as_bytes(sess.run(next_element)))
+      self.assertItemsEqual(full_filenames, produced_filenames)
+
+      with self.assertRaises(errors.OutOfRangeError):
+        sess.run(itr.get_next())
+
+  def testFileMiddles(self):
+    filenames = ['a.txt', 'b.py', 'c.pyc']
+    self._touchTempFiles(filenames)
+
+    filename_placeholder = array_ops.placeholder(dtypes.string, shape=[])
+    dataset = MatchingFilesDataset(filename_placeholder)
+
+    with self.cached_session() as sess:
+      itr = iterator_ops.Iterator.from_structure(dataset.output_types)
+      init_op = itr.make_initializer(dataset)
+      next_element = itr.get_next()
+      sess.run(
+        init_op,
+        feed_dict={filename_placeholder: path.join(self.tmp_dir, '*.py*')})
+
+      full_filenames = []
+      produced_filenames = []
+      for filename in filenames[1:]:
+        full_filenames.append(
+          compat.as_bytes(path.join(self.tmp_dir, filename)))
+        produced_filenames.append(compat.as_bytes(sess.run(next_element)))
+
+      self.assertItemsEqual(full_filenames, produced_filenames)
+
+      with self.assertRaises(errors.OutOfRangeError):
+        sess.run(itr.get_next())
+
```

Comment:
Can you also add some correctness tests for files nested in directories? I know you test this in `testPerformance`, but this should be tested independently of performance and I also later suggest to convert `testPerformance` to a benchmark.

Reply:
Sure, have added a test for files nested in directories. The structure of nested directories is as below:
```
- /tmp_dir/
- /tmp_dir/0
- /tmp_dir/0/a.txt
- /tmp_dir/0/b.py
- /tmp_dir/0/c.pyc
- /tmp_dir/0/0
- /tmp_dir/0/0/a.txt
- /tmp_dir/0/0/b.py
- /tmp_dir/0/0/c.pyc
- /tmp_dir/0/0/1
- /tmp_dir/0/0/1/a.txt
- /tmp_dir/0/0/1/b.py
- /tmp_dir/0/0/1/c.pyc
- /tmp_dir/0/0/1/2
- /tmp_dir/0/0/1/2/a.txt
- /tmp_dir/0/0/1/2/b.py
- /tmp_dir/0/0/1/2/c.pyc
- /tmp_dir/0/0/1/2/3
- /tmp_dir/0/0/1/2/3/a.txt
- /tmp_dir/0/0/1/2/3/b.py
- /tmp_dir/0/0/1/2/3/c.pyc
.
.
.
- /tmp_dir/0/....../depth-1/
- /tmp_dir/0/....../depth-1/a.txt
- /tmp_dir/0/....../depth-1/b.py
- /tmp_dir/0/....../depth-1/c.pyc
.
.
.
- /tmp_dir/width-1
.
.
.
- /tmp_dir/width-1/....../depth-1/
- /tmp_dir/width-1/....../depth-1/a.txt
- /tmp_dir/width-1/....../depth-1/b.py
- /tmp_dir/width-1/....../depth-1/c.pyc
```